File: rust/src/src/rag_manager.rs
// src/rag_manager.rs  <- src/rag_manager.py
//! A thin wrapper around `VectorRAG` for backward compatibility.
//!
//! ## Port classification — PORT_VIA_DB (web)
//!
//! The Python `RAGManager` is a backward-compatibility shim: it constructs a
//! `VectorRAG(persist_directory=...)` and delegates every method straight through.
//! This port reproduces that exactly — `RAGManager` owns a `VectorRAG` *by value*
//! and forwards each call.
//!
//! ### Sync -> async
//!
//! The ported `VectorRAG` is async (its constructor does an HTTP embedding
//! dimension probe; `search`/`retrieve`/`add_*`/`index_personal_documents` await
//! the embedding HTTP client). So `RAGManager::new` and the IO-bound delegating
//! methods are `async`; `rebuild_index` / `get_stats` stay sync (they are sync on
//! `VectorRAG`).
//!
//! ### Richer Rust signatures, reconciled
//!
//! The ported `VectorRAG` methods take a couple of extra optional arguments the
//! Python `RAGManager` did not thread through (`search`'s `owner`,
//! `index_personal_documents`'s `file_extensions` / `owner`). The Python shim
//! always called the defaults, so this wrapper passes `None` for them — preserving
//! the Python `RAGManager` behaviour exactly. `search` maps each `SearchHit` to its
//! Python-dict shape via `SearchHit::to_value()` to reproduce the
//! `List[Dict[str, Any]]` return type.
//!
//! ### Status
//!
//! Dead / backward-compatibility code: there is no in-tree caller of `RAGManager`
//! (the live RAG entrypoints are `rag_singleton::get_rag_manager` — disabled — and
//! `tool_index`'s direct `chroma_client` use). Ported faithfully for completeness.
//!
//! Feature gate: `web` (owns a `VectorRAG`, which needs HTTP embeddings + the
//! rusqlite vector store).

use serde_json::Value;

use crate::pylog as logger;
use crate::src::rag_vector::VectorRAG;

/// A manager class that wraps `VectorRAG` for backward compatibility.
///
/// Most methods delegate directly to `VectorRAG` (mirrors the Python `RAGManager`).
pub struct RAGManager {
    /// The wrapped engine (`self.vector_rag`).
    pub vector_rag: VectorRAG,
}

impl RAGManager {
    /// `RAGManager.__init__(self, persist_directory="data/chroma")`.
    ///
    /// Builds the wrapped `VectorRAG`. Async because the `VectorRAG` constructor
    /// runs `_initialize_system` (an HTTP embedding dimension probe). The Python
    /// default `persist_directory` is `"data/chroma"`; callers pass it explicitly
    /// here.
    pub async fn new(persist_directory: &str) -> Self {
        let vector_rag = VectorRAG::new(persist_directory).await;
        logger::info("RAGManager initialized as wrapper for VectorRAG");
        RAGManager { vector_rag }
    }

    /// `search(self, query, k=5) -> List[Dict[str, Any]]` — delegates to VectorRAG.
    ///
    /// The Python shim called `self.vector_rag.search(query, k)` (no `owner`); we
    /// pass `owner=None`, then render each `SearchHit` as its Python dict.
    pub async fn search(&self, query: &str, k: i64) -> Vec<Value> {
        self.vector_rag
            .search(query, k, None)
            .await
            .iter()
            .map(|hit| hit.to_value())
            .collect()
    }

    /// `index_personal_documents(self, directory) -> dict` — delegates to VectorRAG.
    ///
    /// The Python shim passed only `directory`; we pass `file_extensions=None` /
    /// `owner=None` (the `VectorRAG` defaults the Python shim relied on).
    pub async fn index_personal_documents(&self, directory: &str) -> Value {
        self.vector_rag
            .index_personal_documents(directory, None, None)
            .await
    }

    /// `retrieve(self, query, k=5) -> List[str]` — delegates to VectorRAG.
    pub async fn retrieve(&self, query: &str, k: i64) -> Vec<String> {
        self.vector_rag.retrieve(query, k).await
    }

    /// `rebuild_index(self) -> bool` — delegates to VectorRAG.
    ///
    /// `&mut self` because `VectorRAG::rebuild_index` mutates `healthy`/the
    /// collection handle on a clean rebuild.
    pub fn rebuild_index(&mut self) -> bool {
        self.vector_rag.rebuild_index()
    }

    /// `get_stats(self) -> Dict[str, Any]` — delegates to VectorRAG.
    pub fn get_stats(&self) -> Value {
        self.vector_rag.get_stats()
    }

    /// `add_document(self, text, metadata) -> bool` — delegates to VectorRAG.
    pub async fn add_document(&self, text: &str, metadata: &Value) -> bool {
        self.vector_rag.add_document(text, metadata).await
    }

    /// `add_documents_batch(self, docs) -> Dict[str, Any]` — delegates to VectorRAG.
    ///
    /// `docs` is the Python `List[tuple]` of `(text, metadata)` pairs.
    pub async fn add_documents_batch(&self, docs: &[(String, Value)]) -> Value {
        self.vector_rag.add_documents_batch(docs).await
    }
}

File: src/rag_manager.py
"""RAG manager wrapper around VectorRAG."""
import logging
import os
from typing import Any, Dict, List, Optional

from src.constants import CHROMA_DIR

try:
    from rag_vector import VectorRAG
except ImportError:
    try:
        from .rag_vector import VectorRAG
    except ImportError:
        from src.rag_vector import VectorRAG

logger = logging.getLogger(__name__)


class RAGManager:
    """A manager class that wraps VectorRAG for backward compatibility."""

    def __init__(self, persist_directory: str = CHROMA_DIR):
        self.vector_rag = VectorRAG(persist_directory=persist_directory)
        logger.info("RAGManager initialized as wrapper for VectorRAG")

    def search(self, query: str, k: int = 5, owner: Optional[str] = None) -> List[Dict[str, Any]]:
        """Search for documents, optionally filtering by owner."""
        return self.vector_rag.search(query, k, owner=owner)

    def index_personal_documents(
        self,
        directory: str,
        file_extensions: Optional[set] = None,
        owner: Optional[str] = None,
    ) -> Dict[str, Any]:
        return self.vector_rag.index_personal_documents(
            directory,
            file_extensions=file_extensions,
            owner=owner,
        )

    def retrieve(self, query: str, k: int = 5, owner: Optional[str] = None) -> List[str]:
        if owner is None:
            return self.vector_rag.retrieve(query, k)
        return [r["document"] for r in self.vector_rag.search(query, k, owner=owner)]

    def remove_directory(self, directory: str, owner: Optional[str] = None) -> Dict[str, Any]:
        """Remove indexed chunks under a directory, optionally only for one owner."""
        if owner is None:
            return self.vector_rag.remove_directory(directory)

        if not self.vector_rag.healthy:
            return {"success": False, "message": "Collection not initialized"}

        directory = os.path.abspath(directory)
        owner = (owner or "").strip().lower() or None

        try:
            removed_ids = set()
            collections_for_delete = getattr(self.vector_rag, "_collections_for_delete", None)
            if not callable(collections_for_delete):
                return {"success": False, "message": "VectorRAG does not expose collection deletion"}

            for _lane_name, collection in collections_for_delete():
                results = collection.get(include=["metadatas"])
                ids = [
                    results["ids"][i]
                    for i, meta in enumerate(results["metadatas"])
                    if isinstance(meta, dict)
                    and isinstance(meta.get("source"), str)
                    and (meta["source"] == directory or meta["source"].startswith(directory + os.sep))
                    and str(meta.get("owner", "")).strip().lower() == owner
                ]
                if ids:
                    collection.delete(ids=ids)
                    removed_ids.update(ids)

            if not removed_ids:
                return {"success": True, "removed_count": 0, "message": "No docs found"}

            count = len(removed_ids)
            logger.info("Removed %s chunks from %s for owner=%s", count, directory, owner)
            return {"success": True, "removed_count": count, "message": f"Removed {count} chunks"}
        except Exception as exc:
            logger.error("remove_directory %s owner=%s: %s", directory, owner, exc)
            return {"success": False, "message": str(exc)}

    def reindex_directory(
        self,
        directory: str,
        file_extensions: Optional[set] = None,
        owner: Optional[str] = None,
    ) -> Dict[str, Any]:
        remove_result = self.remove_directory(directory, owner=owner)
        if not remove_result.get("success"):
            return remove_result
        index_result = self.index_personal_documents(directory, file_extensions, owner=owner)
        return {
            "success": index_result.get("success", False),
            "message": (
                f"Re-index for {directory}: removed {remove_result.get('removed_count', 0)}, "
                f"{index_result.get('message', '')}"
            ),
            "removed_count": remove_result.get("removed_count", 0),
            "indexed_count": index_result.get("indexed_count", 0),
            "failed_count": index_result.get("failed_count", 0),
        }

    def rebuild_index(self) -> bool:
        return self.vector_rag.rebuild_index()

    def get_stats(self) -> Dict[str, Any]:
        return self.vector_rag.get_stats()

    def add_document(self, text: str, metadata: Dict[str, Any]) -> bool:
        return self.vector_rag.add_document(text, metadata)

    def add_documents_batch(self, docs: List[tuple]) -> Dict[str, Any]:
        return self.vector_rag.add_documents_batch(docs)

    def delete_by_source(self, source: str) -> int:
        return self.vector_rag.delete_by_source(source)

    def rename_owner(
        self,
        old_owner: str,
        new_owner: str,
        *,
        path_map: Optional[Dict[str, str]] = None,
        path_prefixes: Optional[List[tuple]] = None,
    ) -> Dict[str, Any]:
        return self.vector_rag.rename_owner(
            old_owner,
            new_owner,
            path_map=path_map,
            path_prefixes=path_prefixes,
        )

    def _split_into_chunks(self, *args, **kwargs):
        return self.vector_rag._split_into_chunks(*args, **kwargs)
